src/std/net/socket/server-test.ss

```diff
@@ -0,0 +1,84 @@
+;;; -*- Gerbil -*-
+;;; (C) vyzo at hackzen.org
+;;; Socket server tests
+
+(import :gerbil/gambit/threads
+        :std/misc/threads
+        :std/net/socket
+        :std/net/socket/base
+        :std/os/socket
+        :std/error
+        :std/sugar
+        :std/test)
+(export socket-server-test)
+
+(def socket-server-test
+  (test-suite "test :std/net/socket"
+
+    (def server-address "localhost:4242")
+
+    (def (echo sock)
+      (def buf (make-u8vector 256))
+      (try
+       (let lp ()
+         (def len (ssocket-recv sock buf))
+         (if (= len 0)
+           (begin
+             (ssocket-close sock)
+             (values))
+           (begin
+             (ssocket-send sock buf 0 len)
+             (lp))))
+       (catch (io-error? exn)
+         (values))))
+
+    (def (run-echo-server address)
+      (def sock (ssocket-listen address))
+      (let lp ()
+        (spawn echo (ssocket-accept sock))
+        (lp)))
+
+    (def (start-echo-server! address)
+      (parameterize ((current-socket-server #f))
+        (start-socket-server!)
+        (let (echo-server (spawn/group 'echo-server run-echo-server address))
+          (values (current-socket-server) echo-server))))
+
+    (def (stop-echo-server! server)
+      (let (tgroup (thread-thread-group server))
+        (thread-group-kill! tgroup)))
+
+    (def (receive-data sock)
```

Comment:
i would suggest using a combined `(send+receive-data sock txt)` so that you can use `ssocket-send-all` and `ssocket-recv-all` to avoid issues with incomplete read/writes.